WordPress/gutenberg
# Blocks: Support block nesting (Add Columns block, InnerBlocks component)

Closes #2995 
Closes #428
Partially implements #219
~Cherry-picks 1efbb9faf02b12a64759fddc7027e5bfa089ecfc and 2bd32f8cf0acb0dcbf183f8edae348b32ab1e9b3 from #3743~
Related: #3705, #2743
~Partially reverts #2806 (temporarily)~

This pull request seeks to explore support for generalized nested blocks, implementing a `InnerBlocks` component which effectively behaves as a sub-editor block listing (maintaining its own separate editor state). A block's children are available via its `block.innerBlocks` property, and can be accessed from a component's `edit` and `save` implementations (`props.innerBlocks`), and updated via `edit`'s `setInnerBlocks` prop.

Included is a new Columns block which behaves similarly to the Text Columns block, except that it allows insertion of blocks within the rendered columns.

![Nesting](https://user-images.githubusercontent.com/1779930/33456167-8f3efe96-d5ec-11e7-8177-5304c436ec3b.png)

**What has been implemented:**

- Block parsing, creation, validation, and serialization of nested content
- Editing block children, including insertion, rearrangement, toolbar and inspector controls

**What doesn't work so well yet:**

- Front-end display
- Reconciliation of selected blocks, particularly in rendering of toolbar and inspector slot contents
- Reducing columns trims content instead of collapsing
- General polish and UX behaviors
- Tests and such

__Testing instructions:__

Verify that you can insert a new Columns block, and within it, insert, rearrange, edit, and delete child blocks. Verify that saving the post and refreshing the editor respects all changes which were made.
